xcode/Subconscious/Shared/Services/AddressBookService.swift

```diff
@@ -68,37 +69,36 @@ extension AddressBookError: LocalizedError {
 /// An AddressBook can wrap any Sphere and provide a higher-level interface to manage petnames.
 actor AddressBook<Sphere: SphereProtocol> {
     private var sphere: Sphere
-    private var addressBook: [AddressBookEntry]?
+    private var cache: [Petname:AddressBookEntry] = [:]
     
     /// Logger cannot be static because actor is generic
     private let logger = Logger(
         subsystem: Config.default.rdns,
         category: "AddressBookService"
     )
     
-    init(sphere: Sphere, addressBook: [AddressBookEntry]? = nil) {
+    init(sphere: Sphere, addressBook: [Petname:AddressBookEntry] = [:]) {
         self.sphere = sphere
-        self.addressBook = addressBook
-    }
-    
-    func invalidateCache() {
-        addressBook = nil
+        self.cache = addressBook
     }
     
     /// Get the full list of entries in the address book.
-    /// This is cached after the first use unless requested using refetch.
     /// If an error occurs producing the entries the resulting list will be empty.
-    func listEntries(
-        refetch: Bool = false
-    ) async throws -> [AddressBookEntry] {
-        if let addressBook = addressBook {
-            if !refetch {
-                return addressBook
-            }
-        }
-        var addressBook: [AddressBookEntry] = []
+    func listEntries() async throws -> [AddressBookEntry] {
         let petnames = try await sphere.listPetnames()
+        let version = try await sphere.version()
+        
+        var entries: [AddressBookEntry] = []
+        
         for petname in petnames {
+            // Re-use the cached result if the version matches
+            if let entry = self.cache[petname],
+               version == entry.version {
+                entries.append(entry)
+                
+                continue
+            }
+            
```

Comment:
It seems we are patching the cache entry-by-entry.

A more straightforward approach may be to check if the version differs from the cached version, and if so, wholesale replace the petname entries with those we get back from `sphere.listPetnames()`.

We have to iterate over the full list of petnames anyway, so this should be the same time complexity. Additionally, replacing the cache has the advantage of being able to exit early if the sphere versions match.

Reply:
Hm, I didn't want to pay the cost of rebuilding the entire cache in one hit if we didn't need to but I suppose that's a negligible amount of work in practice.

The address book differs slightly from the user profile cache though in that we never fetch an item directly from the cache - we always list it. That has me leaning towards your suggestion for the address bool but leaving the user profile cache as-is? I'll do another pass on Monday.